changelog.py

```diff
@@ -0,0 +1,168 @@
+
+    def _slug(self, s: str) -> str:
+        """Return a version of the string that's easy to compare."""
+        return re.sub(r"[\s_-]", "", s.casefold())
+
+    def _previous_release(self, version: str):
+        """Get the release previous to the current one (according to SemVer)."""
+        cur_ver = semver.parse_version_info(version[1:])
+        prev_ver = semver.parse_version_info("0.0.0")
+        prev_rel = None
+        for tag in self._repo.tags.list(all=True):
+            if not tag.name.startswith("v"):
+                continue
+            try:
+                ver = semver.parse_version_info(tag.name[1:])
+            except ValueError:
+                continue
+            if ver.prerelease or ver.build:
+                continue
+            # Get the highest version that is not greater than the current one.
+            # That means if we're creating a backport v1.1, an already existing v2.0,
+            # despite being newer than v1.0, will not be selected.
+            if ver < cur_ver and ver > prev_ver:
+                prev_rel = tag
+                prev_ver = ver
+        return prev_rel
+
+    def _issues(self, start: datetime, end: datetime):
+        """Collect issues that were closed in the interval."""
+        issues = []
+        for x in self._repo.issues.list(state="closed", updated_after=start, all=True):
+            # If a previous release's last commit closed an issue, then that issue
+            # should be included in the previous release's changelog and not this one.
+            # The interval includes the endpoint for this same reason.
+            closed_at = parser.parse(x.closed_at)
+            if closed_at <= start or closed_at > end:
+                continue
+            if self._ignore.intersection(self._slug(label) for label in x.labels):
+                continue
+            else:
+                issues.append(x)
+        issues.reverse()  # Sort in chronological order.
+        return issues
+
+    def _merge_requests(self, start: datetime, end: datetime):
+        """Collect merge requests in the interval."""
+        merge_requests = []
+        for x in self._repo.mergerequests.list(state="merged", updated_after=start, all=True):
+            merged_at = parser.parse(x.merged_at)
+            if merged_at <= start or merged_at > end:
+                continue
+            if self._ignore.intersection(self._slug(label) for label in x.labels):
+                continue
+            else:
+                merge_requests.append(x)
+        merge_requests.reverse()  # Sort in chronological order.
+        return merge_requests
+
+    def _format_user(self, user) -> Dict[str, object]:
+        """Format a user for the template."""
+        if user is None:
+            return None
+
+        return {
+            "name": user['name'],
+            "url": user['web_url'],
+            "username": user['username'],
+        }
+
+    def _format_issue(self, issue) -> Dict[str, object]:
```

Comment:
if you want to type hint the gitlab objects on the functions you can do something like: `def _format_issue(self, issue: gitlab.v4.objects.ProjectIssueManager) -> Dict[str, object]:` and if thats too long, at the top you could `from gitlab.v4.objects import ProjectIssueManager` and then `def _format_issue(self, issue: ProjectIssueManager) -> Dict[str, object]:` 

That might not be the right object, but just for example